### ISSUE TITLE
Spec writable protocol handlers

### ISSUE THREAD

**Issue by RangerMauve** [ipfs/specs#335]
As per @lidel's request here's an issue to track the discussion we're having on standardizing writability in IPFS protocol handlers.

The tl;dr, we've got `ipfs://` and `ipns://` protocol handlers in browsers like Agregore/Brave/Capyloon and most of them focus on what it means to `GET` some data from the protocol.

In Agregore's case, we've also had the ability to `PUT` or `POST` data to IPFS.

Initially this was done via the `POST` verb which would take some data, upload it to IPFS, then return the uploaded IPFS URL in the response body with a 200 status code.

Earlier this year, during the Agregore Mobile devgrant, we forked Kubo's [writable gateway](https://github.com/ipfs/kubo/blob/master/core/corehttp/gateway_handler.go#L440) in order to get IPFS protocol handlers on Android with a Chromium fork. Part of this effort involved writing [a spec](https://github.com/AgregoreWeb/agregore-ipfs-daemon/blob/main/spec.md) based on Kubo's writable APIs.

The main thing that changed was that Kubo was using the [201](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/201) status code and placing the resulting CID in the `Location` header instead of into the response body. Initially I was aprehensive and wanted to use what we already had, but @makeworld-the-better-one convinced me using solid reasoning which I no longer remember that we should switch over.

Most recently, @fabricedesre was talking about wanting to move back to the 200+body approach for showing result URLs. It seems like this might be the apoproach he'd like to take for Capyloon which is now embedding an Iroh node which could lead to local authoring of data.

Personally, I'd like to keep the 201+location method if possible since I have code which relies on that already, but it's not a hard blocker and I'd be willing to update stuff if it means we can get writability into other browsers sooner.

This also relates to discussions we're having for [ipld](https://github.com/ipfs/specs/pull/293) protocol handlers where we need to specify what an IPLD POST/PATCH should yield.

Ideally we will book a time during Labweek 2022 in Lisbon to sit down and talk about this stuff to figure out what approach we align on.

**Comment by ikreymer** [comment]
If the primary application is access to a local and/or embedded ipfs mode by a webpage in a browser, then I think the design needs to include the the security / permissions model for a website pinning data to a local node.
Otherwise, no matter what the POST/PUT API is, a browser vendor can't be expected to just allow any website to connect to a localhost port and POST/PUT unlimited data.

Some ideas include extending `navigator.permissions.query` and then perhaps another api to get an auth token that is passed to all APIs. A next step would also be doing something with the quota system. All of this is a bit more tricky, but probably worth figuring out to be able to support true p2p ipfs access in browsers. (Maybe this should be its own issue, though :)

**Comment by RangerMauve** [comment]
Still into this. Any clue how I can book time at PL Summit? Wednesday I'm doing some IPLD roadmapping, but I'm free outside of that.

**Comment by RangerMauve** [comment]
Looks like some time during IPFS Camp would be good. Maybe a chat during breakfast before one of the days.

**Comment by RangerMauve** [comment]
Might do Friday?

**Comment by RangerMauve** [comment]
Notes from meetup at Lisbon:

- We should have content-type as part of the data when uploading @darobin Should be strict with top level content type and how it's processed
- Posgt to ipfs://localhost to upload a binary or a formdata
  - Might want to bikeshed localhost
- PUT to `ipfs://CID/path` to add a file or upload a folder
- Would be useful to think about authorization for the gateway use case @gozala has thoughts from web3.storage
- Would be nice to spec CAR file upload soon after this initial thing
- Using redirects in the response can lead to a bunch of issues, should return the URL itself
- 201 is useful in that it's semantically correct, and in a form submission it doesn't navigate away.
- Should have the PUT over data be a separate spec, then stuff like upload APIs on gateways won't need to bother with this

**Comment by RangerMauve** [comment]
One thing that would be ueful for IPFS is to import public keys from WebCrypto

User agent (browser) can decide to provide users with a UI to manage key access

Capolyoon is down to experiment with soem initial IPFS//IPNS upload APIs

**Comment by RangerMauve** [comment]
Will work on an initial spec PR for protocol handlers and go from there.

Might be useful to have stuff about where to save things for the gateway use case

**Comment by RangerMauve** [comment]
@darobin would you be down to have a call some time soon to start working on the spec together? Maybe we could have an agenda item on the browsers and standards call? cc @autonome would that work? Would also be good to have Fabrice in there

**Comment by darobin** [comment]
Sure thing. I've been going through the existing specs and also looking at what Kubo does. I think having the discussion with the whole gang might be the best option.
